src/unstract/sdk/utils/common_utils.py

```diff
@@ -54,3 +55,51 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return decorator
+
+
+def capture_metrics(func):
+    """Decorator to capture metrics at the start and end of a function."""
+
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        # Ensure the required attributes exist; if not,
+        # execute the function and return its result
+        if not all(
+            hasattr(self, attr) for attr in ["_run_id", "_capture_metrics", "_metrics"]
+        ):
+            return func(self, *args, **kwargs)
+
+        # Check if run_id exists and if metrics should be captured
+        metrics_mixin = None
+        time_taken_key = MetricsMixin.TIME_TAKEN_KEY
+        if self._run_id and self._capture_metrics:
+            metrics_mixin = MetricsMixin(run_id=self._run_id)
+
+        try:
+            result = func(self, *args, **kwargs)
+        finally:
+            # If metrics are being captured, collect and assign them at the end
+            if metrics_mixin:
+                new_metrics = metrics_mixin.collect_metrics()
+
+                # If time_taken(s) exists in both self._metrics and new_metrics, sum it
+                if (
+                    self._metrics
+                    and time_taken_key in self._metrics
+                    and time_taken_key in new_metrics
+                ):
+                    time_taken_self = self._metrics.get(time_taken_key)
+                    time_taken_new = new_metrics.get(time_taken_key)
+
+                    # Only sum if both are valid
+                    if time_taken_self and time_taken_new:
+                        self._metrics[time_taken_key] = time_taken_self + time_taken_new
+                    else:
+                        self._metrics[time_taken_key] = None
```

Comment:
NIT: Currently this might be fine - but @Deepak-Kesavan do you think its okay to set this to `None` if one of the `time_taken_*` is not available?

Reply:
Yes @chandrasekharan-zipstack . I added it as fallback if key is not available in redis. Let's see how this works and I believe it will never hit this case.